src/group_theory/commensurable.lean

```diff
@@ -0,0 +1,119 @@
+
+lemma comm {H K : subgroup G} : commensurable H K ↔ commensurable K H := and.comm
+
+@[symm] lemma symm {H K : subgroup G} : commensurable H K → commensurable K H := and.symm
+
+@[trans] lemma trans {H K L : subgroup G} (hhk : commensurable H K ) (hkl : commensurable K L) :
+  commensurable H L :=
+⟨subgroup.relindex_ne_zero_trans H K L hhk.1 hkl.1,
+  subgroup.relindex_ne_zero_trans L K H hkl.2 hhk.2⟩
+
+lemma equivalence : equivalence (@commensurable G _) :=
+⟨commensurable.refl, λ _ _, commensurable.symm, λ _ _ _, commensurable.trans⟩
+
+/--Equivalence of `K/H ⊓ K` with `gKg⁻¹/gHg⁻¹ ⊓ gKg⁻¹`-/
+def  quot_conj_equiv (H K : subgroup G) (g : conj_act G) :
+  quotient_group.quotient (H.subgroup_of K) ≃
+  quotient_group.quotient ((g • H).subgroup_of (g • K)) :=
+quotient.congr (K.equiv_smul g).to_equiv (λ a b, by rw [←quotient.eq', ←quotient.eq',
+  quotient_group.eq', quotient_group.eq', subgroup.mem_subgroup_of, subgroup.mem_subgroup_of,
+  mul_equiv.coe_to_equiv, ←mul_equiv.map_inv, ←mul_equiv.map_mul,
+  subgroup.equiv_smul_apply_coe, subgroup.smul_mem_pointwise_smul_iff])
+
+lemma commensurable_conj {H K : subgroup G} (g : conj_act G) :
+   commensurable H K ↔ commensurable (g • H) (g • K) :=
+and_congr (not_iff_not.mpr (eq.congr_left (cardinal.to_nat_congr (quot_conj_equiv H K g))))
+  (not_iff_not.mpr (eq.congr_left (cardinal.to_nat_congr (quot_conj_equiv K H g))))
+
+lemma commensurable_inv (H : subgroup G) (g : conj_act G) :
+  commensurable (g • H) H ↔ commensurable H (g⁻¹ • H) :=
+ by rw [commensurable_conj, inv_smul_smul]
+
+/--For `H` a subgroup of `G`, this is the subgroup of all elements `g : conj_aut G`
+such that `commensurable ( g • H) H`   -/
+
+def commensurator' (H : subgroup G) : subgroup (conj_act G) :=
+{ carrier := {g : conj_act G | commensurable (g • H) H},
+  one_mem' := by rw [set.mem_set_of_eq, one_smul],
+  mul_mem' := λ a b ha hb, by
+  { rw [set.mem_set_of_eq, mul_smul],
+    exact trans ((commensurable_conj a).mp hb) ha },
+  inv_mem' := λ a ha, by rwa [set.mem_set_of_eq, comm, ←commensurable_inv] }
+
+/--For `H` a subgroup of `G`, this is the subgroup of all elements `g : G`
+such that `commensurable ( g H g⁻¹) H`   -/
+
+def commensurator (H : subgroup G) : subgroup G :=
+  (commensurator' H).map (conj_act.of_conj_act.to_monoid_hom)
+
+@[simp]
+lemma commensurator'_mem_iff (H : subgroup G) (g : conj_act G) :
+  g ∈ (commensurator' H) ↔ commensurable (g • H) H := iff.rfl
+
+@[simp]
+lemma commensurator_mem_iff (H : subgroup G) (g : G) :
+  g ∈ (commensurator H) ↔ commensurable (conj_act.to_conj_act g • H) H :=
+begin
+rw commensurator,
+simp only [exists_prop, mul_equiv.coe_to_monoid_hom, commensurator'_mem_iff, subgroup.mem_map],
+split,
+intro h,
+obtain ⟨x, hx⟩ := h,
+rw ← hx.2,
+simp only [conj_act.to_conj_act_of_conj_act],
+apply hx.1,
+intro h,
+use conj_act.to_conj_act g,
+simp only [h, conj_act.of_conj_act_to_conj_act, eq_self_iff_true, and_self],
+end
```

Comment:
```suggestion
def commensurator (H : subgroup G) : subgroup G :=
(commensurator' H).comap (conj_act.to_conj_act.to_monoid_hom)

@[simp] lemma commensurator'_mem_iff (H : subgroup G) (g : conj_act G) :
  g ∈ (commensurator' H) ↔ commensurable (g • H) H := iff.rfl

@[simp] lemma commensurator_mem_iff (H : subgroup G) (g : G) :
  g ∈ (commensurator H) ↔ commensurable (conj_act.to_conj_act g • H) H := iff.rfl
```

Reply:
Oh haha so the `reverse` definition simplifies nicely but not the one I tried. Thanks this is much nicer now!

Reply:
As a general rule of thumb, `subgroup.comap` is much nicer to work with than `subgroup.map`.